pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go

```diff
@@ -0,0 +1,116 @@
+// This file defines a custom key to implement the logic for rego rule:
+
+package jsonSchemaValidator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+const RegoDefinitionCustomKey = "regoDefinition"
+
+var regoCodeToEval = RegoDefinition{}
+var mainRegoPackage = ""
+var mainModuleFileName = "main.rego"
+var regoFunctionEntryPoint = "violation"
+
+type CustomKeyRegoDefinitionCompiler struct{}
+
+type CustomKeyRegoDefinitionSchema map[string]interface{}
+
+var CustomKeyRegoRule = jsonschema.MustCompileString("customKeyRegoDefinition.json", `{
+	"properties" : {
+		"regoDefinition": {
+			"type": "object"
+		}
+	}
+}`)
+
+func (CustomKeyRegoDefinitionCompiler) Compile(ctx jsonschema.CompilerContext, m map[string]interface{}) (jsonschema.ExtSchema, error) {
+	if customKeyRegoRule, ok := m[RegoDefinitionCustomKey]; ok {
+		customKeyRegoRuleObj, validObject := customKeyRegoRule.(map[string]interface{})
+		if !validObject {
+			return nil, fmt.Errorf("regoDefinition must be an object")
+		}
+
+		b, _ := json.Marshal(customKeyRegoRule)
+		var regoDefinition RegoDefinition
+		err := json.Unmarshal(b, &regoDefinition)
+		if err != nil {
+			return nil, fmt.Errorf("regoDefinition must be an object of type RegoDefinition")
+		}
+
+		regoCodeToEval = regoDefinition
+		return CustomKeyRegoDefinitionSchema(customKeyRegoRuleObj), nil
+	}
+	return nil, nil
+}
+
+type RegoDefinition struct {
+	Libs []string `json:"libs"`
+	Code string   `json:"code"`
+}
+
+func (s CustomKeyRegoDefinitionSchema) Validate(ctx jsonschema.ValidationContext, dataValue interface{}) error {
+	regoCtx := context.Background()
+
+	r := retrieveRegoFromSchema(regoCodeToEval.Code)
+
+	// Create a prepared query that can be evaluated.
+	query, err := r.PrepareForEval(regoCtx)
+	if err != nil {
+		return ctx.Error(RegoDefinitionCustomKey, "can't compile rego code, %s", err.Error())
+	}
+
+	// Execute the prepared query.
+	rs, err := query.Eval(regoCtx, rego.EvalInput(dataValue))
+
+	if err != nil {
+		return ctx.Error(RegoDefinitionCustomKey, "failed to evaluate rego due to %s", err.Error())
+	}
+
+	if len(rs) != 1 || len(rs[0].Expressions) != 1 {
+		return ctx.Error(RegoDefinitionCustomKey, "failed to evaluate rego, unexpected results")
+	} else {
+		resultsValue := (rs[0].Expressions[0].Value).([]interface{})
+		value, ok := resultsValue[0].(bool)
+		if ok {
+			if value {
+				return ctx.Error(RegoDefinitionCustomKey, "values in data value %v do not match", rs[0].Expressions[0].Value)
+			}
+			return nil
+		} else {
+			return ctx.Error(RegoDefinitionCustomKey, "violation needs to return a boolean")
+		}
+	}
+}
+
+func getPackageFromRegoCode(regoCode string) string {
+	const PACKAGE = "package"
```

Comment:
if the word "package" dose not exist in `regoCode`, it can be a panic error?

<img width="539" alt="image" src="https://user-images.githubusercontent.com/29673827/224688971-e7ccac13-8b04-4249-b50b-99c8f5c545dc.png">

Reply:
I don't think that panicing here will be good because imagine we have a policy with 6 different rules, 5 are ok and 1 is not, instead of evaluating the rule we will panic the policyCheck, feels like a bad experience